team_project_access.go

```diff
@@ -0,0 +1,218 @@
+	// https://jsonapi.org/format/#crud-creating
+	Type string `jsonapi:"primary,team-projects"`
+	// The type of access to grant.
+	Access *TeamProjectAccessType `jsonapi:"attr,access,omitempty"`
+}
+
+// List all team accesses for a given project.
+func (s *teamProjectAccesses) List(ctx context.Context, options *TeamProjectAccessListOptions) (*TeamProjectAccessList, error) {
+	if err := options.valid(); err != nil {
+		return nil, err
+	}
+
+	req, err := s.client.NewRequest("GET", "team-projects", options)
+	if err != nil {
+		return nil, err
+	}
+
+	tpal := &TeamProjectAccessList{}
+	err = req.Do(ctx, tpal)
+	if err != nil {
+		return nil, err
+	}
+
+	return tpal, nil
+}
+
+// Add team access for a project.
+func (s *teamProjectAccesses) Add(ctx context.Context, options TeamProjectAccessAddOptions) (*TeamProjectAccess, error) {
+	if err := options.valid(); err != nil {
+		return nil, err
+	}
+
+	req, err := s.client.NewRequest("POST", "team-projects", &options)
+	if err != nil {
+		return nil, err
+	}
+
+	tpa := &TeamProjectAccess{}
+	err = req.Do(ctx, tpa)
+	if err != nil {
+		return nil, err
+	}
+
+	return tpa, nil
+}
+
+// Read a team project access by its ID.
+func (s *teamProjectAccesses) Read(ctx context.Context, teamProjectAccessID string) (*TeamProjectAccess, error) {
+	if !validStringID(&teamProjectAccessID) {
+		return nil, ErrInvalidTeamProjectAccessID
+	}
+
+	u := fmt.Sprintf("team-projects/%s", url.QueryEscape(teamProjectAccessID))
+	req, err := s.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	tpa := &TeamProjectAccess{}
+	err = req.Do(ctx, tpa)
+	if err != nil {
+		return nil, err
+	}
+
+	return tpa, nil
+}
+
+// Update team access for a project.
+func (s *teamProjectAccesses) Update(ctx context.Context, teamProjectAccessID string, options TeamProjectAccessUpdateOptions) (*TeamProjectAccess, error) {
+	if !validStringID(&teamProjectAccessID) {
+		return nil, ErrInvalidTeamProjectAccessID
+	}
+
+	u := fmt.Sprintf("team-projects/%s", url.QueryEscape(teamProjectAccessID))
+	req, err := s.client.NewRequest("PATCH", u, &options)
+	if err != nil {
+		return nil, err
+	}
+
+	ta := &TeamProjectAccess{}
+	err = req.Do(ctx, ta)
+	if err != nil {
+		return nil, err
+	}
+
+	return ta, err
+}
+
+// Remove team access from a project.
+func (s *teamProjectAccesses) Remove(ctx context.Context, teamProjectAccessID string) error {
+	if !validStringID(&teamProjectAccessID) {
+		return ErrInvalidTeamProjectAccessID
+	}
+
+	u := fmt.Sprintf("team-projects/%s", url.QueryEscape(teamProjectAccessID))
+	req, err := s.client.NewRequest("DELETE", u, nil)
+	if err != nil {
+		return err
+	}
+
+	return req.Do(ctx, nil)
+}
+
+func (o *TeamProjectAccessListOptions) valid() error {
+	if o == nil {
+		return ErrRequiredTeamProjectAccessListOps
+	}
+	if !validString(&o.ProjectID) {
+		return ErrRequiredProjectID
+	}
+	if !validStringID(&o.ProjectID) {
+		return ErrInvalidProjectID
+	}
```

Comment:
validString is a logical subset of validStringID so you can omit it

```suggestion
	if !validStringID(&o.ProjectID) {
		return ErrInvalidProjectID
	}
```